### PR TITLE
fix(build): fail the build on a non-relative import in src/ (#137)

### LINKED ISSUE #137: build.js strips every import line indiscriminately — a node: or bare-specifier import in src/ would ship as an undefined reference in the bundle
Surfaced during review of PR #136 (#105 Biome js-lint gate).

## Problem / Goal

`build.js`'s `strip()` removes **every** `import … from …` line indiscriminately when producing `workflows/pipeline.js`. Nothing distinguishes a relative import (correct — the module body is inlined into the bundle) from a `node:*` or bare-specifier import (wrong — nothing is inlined, so the stripped binding ships as an undefined reference).

The failure path, should such an import ever land in `src/`:

1. Author adds `import { inspect } from 'node:util'` to a `workflows/src/*.js` file.
2. The #105 lint gate passes — `noUndeclaredVariables` sees the binding declared in the src file, and `noRestrictedGlobals` only covers globals, not imports. `pipeline.js` is excluded from lint entirely.
3. `node workflows/build.js` strips the import line; the bundle-fresh check passes (the committed bundle matches the rebuild — both are wrong together).
4. Tests pass unless that exact branch is exercised against the bundle.
5. A live dispatch on that path throws `inspect is not defined` — the same coverage-dependent live-crash class #105 closed for sandbox-absent globals.

Today this is guarded only by prose (the tooling-boundary section in root `AGENTS.md`, reworded in PR #136 review). Per the repo design rule, whatever code can enforce, it must.

## Evidence

- 2026-08-04, PR #136 head `4dde11e`: `grep "^import" workflows/src/*.js` — every specifier is relative (`./…`); zero `node:` or bare specifiers. The guard lands green at HEAD — a ratchet, not a cleanup.
- `workflows/build.js:13-15` legitimately uses `node:fs` / `node:url` / `node:path` — it is build tooling, not shipped runtime, and is not stripped into the bundle.
- The workflow sandbox provides no Node builtins (`workflows/AGENTS.md` sandbox-absent list; `applyChallenges.js:16` records the historical `structuredClone` live-smoke crash of the same class).

## Fix shape (one edit)

In `build.js`, at the point `strip()` matches an import line, assert the captured specifier starts with `./` and throw with the offending file/line otherwise. Build fails loudly at CI's existing `node workflows/build.js && git diff --exit-code workflows/pipeline.js` step — no new job, no new config.

## Verification

Mutation, per the repo rule: add `import { inspect } from 'node:util'` to a file under `workflows/src/`, observe `node workflows/build.js` fail naming the file, revert. Suites green at HEAD. Measurement tier: `suites-only`.

### PR DESCRIPTION
**Pull request by leehopper**
Closes #137. Wave 2 quick win; measurement tier `suites-only`.

## Problem

`build.js`'s `strip()` dropped **every** `import … from …` line indiscriminately. A relative
sibling import is safe to drop — ORDER inlines that module's body into the bundle. A `node:*` or
bare specifier inlines nothing, so the stripped binding ships as an undefined reference.

Every existing gate calls that clean:

- Biome (#105) — `noUndeclaredVariables` sees the binding declared in the src file, and
  `pipeline.js` is excluded from lint entirely.
- `node workflows/build.js && git diff --exit-code workflows/pipeline.js` — the committed bundle
  matches the rebuild; both are wrong together.
- The suites — unless that exact branch runs against the bundle.

It surfaces as a live-dispatch `inspect is not defined`: the same coverage-dependent crash class as
the historical `structuredClone` live-smoke failure, since the sandbox provides no Node builtins.
Guarded until now only by prose in the tooling-boundary section.

## Fix

`unsafeImports(source)` returns a violation for any import line whose specifier is not
`./`-relative, and `build()` throws before emitting, naming file, line, specifier and reason. It
also rejects a **multi-line** import: `strip()`'s regex is single-line, so such a statement would
survive into the bundle verbatim.

No new job and no new config — the failure lands on CI's existing rebuild step.

## Verification

Mutation, per the repo rule:

- Added `import { inspect } from 'node:util'` to `workflows/src/registry.js`; `node
  workflows/build.js` exits 1 with `src/registry.js:1: import { inspect } from 'node:util'` and the
  reason. Reverted; bundle byte-identical.
- Stubbed `unsafeImports` to return no violations: the three negative unit tests go red (13 pass /
  3 fail), the positive ones stay green.

Green at HEAD — `grep "^import" workflows/src/*.js` is already all-relative — so this is a ratchet,
not a cleanup.

Gates run locally, all green: Python 1178 passed @ 92.48% (floor 91.7), bench 537 passed @ 88.45%
(floor 87), JS 637 passed @ 98.59/83.65/97.41 (floors 98/82.3/97), `pre-commit run --all-files`.

## Note on the AGENTS.md ratchet

The `AGENTS_SET_BUDGET_BYTES` ratchet is raised 19_282 → 19_682 in the same commit, with the reason
inline at the constant: the mechanism is now code, but the rule an author needs *before* writing the
import is not — `build.js`'s comment is only read after the failure, and neither the Biome gate nor
the bundle-fresh check reports one.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01HC4cXmMt3DkgBXo5Hgbogn

---

> [!NOTE]
> **Medium Risk**
> Build-time enforcement changes contributor workflow for workflow sources; the `resolvePolicy` shape change could affect any consumer expecting `note`, though tests were updated.
> 
> **Overview**
> **`workflows/build.js` now fails at build time** on imports that `strip()` cannot safely remove: non-`./` specifiers (e.g. `node:util`), side-effect imports, and multi-line imports. New **`unsafeImports`** / **`checkUnsafeImports`** run as step 0 before the bundle is emitted, with unit tests covering the throw path and edge cases.
> 
> Contributor docs and CI alignment: **PR checklist** adds `claude plugin validate .`; **CONTRIBUTING** documents which required checks are locally reproducible vs CI-only (Biome lint, PR title). **`tests/test_contribution_surface.py`** enforces that mapping against the PR template and CONTRIBUTING.
> 
> **`resolvePolicy`** in `registry.js` (and the regenerated bundle) **drops the `note` field**; override provenance is documented as living on `resolvedPolicy.subagentModel` instead. **`AGENTS_SET_BUDGET_BYTES`** is raised to document legal import shapes in agent instructions.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8a038e565924cc621c7771b082a4639f3eefd004. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: workflows/build.js [hunk 1]
```diff
@@ -85,6 +85,40 @@ function present() {
...
+      bad.push({
+        line: i + 1, text: line.trim(), specifier: null,
+        reason: 'no import specifier on this line — a multi-line import survives stripping and ships into the bundle verbatim; put the statement on one line',
+      });
+    } else if (!specifier.startsWith('./')) {
+      bad.push({
+        line: i + 1, text: line.trim(), specifier,
+        reason: `specifier '${specifier}' is not relative to src/ — nothing is inlined for it, so stripping the line ships an undefined reference; inline the value into src/ instead`,
+      });
+    }
```
[the remark this answers is not in the document]
Comment (leehopper) [reply workflows/build.js:120]: Valid, and fixed in 73294f5 — thank you. `strip()` matches only `import … from …`, so `import './sideEffect.js';` was never dropped and shipped into the bundle verbatim, despite passing the new `./` check.

Reproduced against 369f9af: adding `import './filterFindings.js';` to `workflows/src/registry.js` built clean (exit 0) and emitted the line into `workflows/pipeline.js`.

The two branches turned out to be one defect seen twice — the line ships as written — so they now share a single condition and reason: the `from` clause must be on the same line. `IMPORT_SPECIFIER` no longer matches a side-effect import's specifier, which drops that whole class into the existing branch rather than adding a third. Nothing in `src/` needs a side-effect import anyway: ORDER already concatenates every module body.

The positive unit test no longer asserts the wrong acceptance, a new test covers the relative side-effect form, and the `workflows/AGENTS.md` bullet now states both clauses.

File: workflows/build.js [hunk 2]
```diff
@@ -120,18 +157,30 @@ export function detectTopLevelCollisions(bundleText) {
...
+
+  // 0) Fail on any import strip() cannot safely drop (see unsafeImports).
+  const unsafe = files.flatMap((file) =>
+    unsafeImports(sources.get(file)).map((v) => ({ ...v, file })));
+  if (unsafe.length) {
+    throw new Error(
+      `build.js: unsafe import(s) in workflows/src — only single-line './sibling.js' imports may be stripped:\n`
+        + unsafe.map((v) => `  src/${v.file}:${v.line}: ${v.text}\n    ${v.reason}`).join('\n'),
+    );
+  }
```
Comment (leehopper) [inline workflows/build.js]: **💡 [LOW] build()'s unsafe-import guard is never exercised through its throw path**

The new step 0 in build() (lines 162-170) collects unsafeImports() violations across every src file, spreads a `file` property onto each, and throws a formatted Error listing them; no test ever puts an actual unsafe import in a source tree and calls build() to confirm it throws, so the wiring around the well-tested pure unsafeImports() function is unverified.

**Suggested fix:**
Add a test that writes (or otherwise stages) a temporary src-like fixture through the same code path — e.g. refactor build() to accept an injectable file-source map for testing, or add a small helper that exercises the `unsafe.length` branch directly by calling the same flatMap/throw logic with a fixture map — and assert the thrown message contains the file name, line number and reason for a synthetic violation spanning two files, so the `{ ...v, file }` spread and the `if (unsafe.length)` gate are both actually verified.

**Cited rule:** AGENTS.md: "A regression test must fail against the bug it names. Verify that by mutating the implementation and watching it go red"

**Review by leehopper — COMMENTED** [review]
Code Gauntlet review: 1 finding survived verification (suggestion-tagged, test coverage).

---
Generated by code-gauntlet | Reviewed up to: 73294f5d1d83e4c7cd66339d5314fd7cfea7c827
